mcs/resource_mcs_bs_volume.go

```diff
@@ -0,0 +1,325 @@
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+const (
+	bsVolumeCreateTimeout = 30 * time.Minute
+	bsVolumeDelay         = 10 * time.Second
+	bsVolumeMinTimeout    = 3 * time.Second
+)
+
+var (
+	bsVolumeStatusBuild    = "creating"
+	bsVolumeStatusActive   = "available"
+	bsVolumeStatusShutdown = "deleting"
+	bsVolumeStatusDeleted  = "deleted"
+)
+
+func resourceBlockStorageVolume() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceBlockStorageVolumeCreate,
+		Read:   resourceBlockStorageVolumeRead,
+		Update: resourceBlockStorageVolumeUpdate,
+		Delete: resourceBlockStorageVolumeDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(bsVolumeCreateTimeout),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+
+			"size": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: false,
+			},
+
+			"volume_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+
+			"availability_zone": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+
+			"metadata": {
+				Type: schema.TypeMap,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				ForceNew: false,
+			},
+
+			"snapshot_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"source_volume_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"image_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceBlockStorageVolumeCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(configer)
+	BlockStorageV2Client, err := config.BlockStorageV2Client(getRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating OpenStack block storage client: %s", err)
+	}
+
+	createOpts := &volumeCreateOpts{}
+	createOpts.AvailabilityZone = d.Get("availability_zone").(string)
+	createOpts.VolumeType = d.Get("volume_type").(string)
+	createOpts.Size = d.Get("size").(int)
+
+	if v, ok := d.GetOk("name"); ok {
+		createOpts.Name = v.(string)
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		createOpts.Description = v.(string)
+	}
+
+	rawMetadata := d.Get("metadata").(map[string]interface{})
+	metadata, err := extractBlockStorageMetadataMap(rawMetadata)
+	if err != nil {
+		return err
+	}
+	createOpts.Metadata = metadata
+
+	if v, ok := d.GetOk("snapshot_id"); ok {
+		createOpts.SnapshotID = v.(string)
+	}
+
+	if v, ok := d.GetOk("source_volume_id"); ok {
+		createOpts.SourceVolID = v.(string)
+	}
+
+	if v, ok := d.GetOk("image_id"); ok {
```

Comment:
These three are exlusive and the only one allowed. We can handle this on a cloud side but maybe it's better to reject it on a client side to avoid late runtime confusion?

Reply:
ok